src/BMP.cpp

```
@@ -0,0 +1,139 @@
+
+        // reading color palette
+        if(bmp_dib_header.bit_per_pixel == BIT_PER_PIXEL_8) {
+            if (bmp_dib_header.colors_used > 0) {
+                inp.read((char*)&bmp_color_palette, sizeof(uint8_t) * bmp_dib_header.colors_used);
+            } else if (bmp_dib_header.colors_used == 0) {
+                inp.read((char*)&bmp_color_palette, sizeof(bmp_color_palette));
+            } else {
+                throw std::runtime_error("Unable to find the color palette of the image.");
+            }
+        }
+
+        // Jump to the pixel data location
+        // inp.seekg(bmp_header.offset_pixel_array, inp.beg);
+        bmp_bitMapArray.resize(bmp_dib_header.width * bmp_dib_header.height * bmp_dib_header.bit_per_pixel / SIZE_OF_BYTE);
+
+        // Checking the different row paddings
+        if (bmp_dib_header.width % 4 == 0) {
+            inp.read((char*)bmp_bitMapArray.data(), bmp_bitMapArray.size());
+            bmp_header.file_size += static_cast<uint32_t>(bmp_bitMapArray.size());
+        } else {
+            row_stride = bmp_dib_header.width * bmp_dib_header.bit_per_pixel / SIZE_OF_BYTE;
+            uint32_t new_stride = make_stride_aligned(4);
+            std::vector<uint8_t> padding_row(new_stride - row_stride); // vector of zero at the size of the padding row
+
+            // reading the array data pixels into the data (and the padding_row)
+            for (int i = 0; i < bmp_dib_header.height; ++i) {
+                inp.read((char*)(bmp_bitMapArray.data() + row_stride * i), row_stride);
+                inp.read((char*)padding_row.data(), padding_row.size());
+            }
+            // Adding the bits that we read to the file's size
+            bmp_header.file_size += static_cast<uint32_t>(bmp_bitMapArray.size()) + bmp_dib_header.height * static_cast<uint32_t>(padding_row.size());
+        }
+    }
+    else {
+        throw std::runtime_error("Unable to open the input image file.");
```

`system_error` (C29)
Error is not propagated (C32)